filebeat/beater/filebeat.go

```diff
@@ -96,23 +105,47 @@ func (fb *Filebeat) Run(b *beat.Beat) error {
 	defer spooler.Stop()
 	defer publisherChan.Close()
 
-	err = crawler.Start(registrar.GetStates())
+	err = crawler.Start(registrar.GetStates(), *once)
 	if err != nil {
 		return err
 	}
 	// Stop crawler -> stop prospectors -> stop harvesters
 	defer crawler.Stop()
 
-	// Blocks progressing. As soon as channel is closed, all defer statements come into play
+	// Wait for completion if running once.
+	if *once {
+		go func() {
+			logp.Info("Running filebeat once. Waiting for completion ...")
+			crawler.WaitForCompletion()
+
+			// Flushes spooler as often as needed to send all events to publisher
+			for spooler.Flush() > 0 {
+			}
```

Comment:
we should a a `defer Flush` to the spooler. The spooler was changed to drain the receive channel until empty, but final flush on shutdown is missing.